File: code/class-04-inheritance/InheritanceDemo.java
/**
 * Animal.java  /  Dog.java  /  Cat.java  /  InheritanceDemo.java
 *
 * Class 04 – Inheritance
 *
 * Demonstrates:
 *   - Parent class with protected fields
 *   - Child classes using `extends` and `super`
 *   - Method overriding with @Override
 *   - Calling a superclass method from a subclass
 */

// ===================== Parent class =====================
class Animal {
    protected String name;
    protected int    age;

    public Animal(String name, int age) {
        this.name = name;
        this.age  = age;
    }

    public void eat() {
        System.out.println(name + " is eating.");
    }

    public void sleep() {
        System.out.println(name + " is sleeping.");
    }

    @Override
    public String toString() {
        return "Animal[name=" + name + ", age=" + age + "]";
    }
}

// ===================== Dog (child) =====================
class Dog extends Animal {
    private String breed;

    public Dog(String name, int age, String breed) {
        super(name, age);    // call Animal constructor
        this.breed = breed;
    }

    public void bark() {
        System.out.println(name + " says: Woof!");
    }

    @Override
    public void eat() {
        super.eat();         // reuse parent behaviour …
        System.out.println(name + " wags its tail while eating.");  // … then extend it
    }

    @Override
    public String toString() {
        return "Dog[name=" + name + ", age=" + age + ", breed=" + breed + "]";
    }
}

// ===================== Cat (child) =====================
class Cat extends Animal {
    private boolean isIndoor;

    public Cat(String name, int age, boolean isIndoor) {
        super(name, age);
        this.isIndoor = isIndoor;
    }

    public void purr() {
        System.out.println(name + " purrs contentedly.");
    }

    @Override
    public void eat() {
        System.out.println(name + " eats very delicately.");
    }

    @Override
    public String toString() {
        return "Cat[name=" + name + ", age=" + age + ", indoor=" + isIndoor + "]";
    }
}

// ===================== Demo =====================
public class InheritanceDemo {
    public static void main(String[] args) {
        Dog dog = new Dog("Rex",      3, "Labrador");
        Cat cat = new Cat("Whiskers", 5, true);

        System.out.println(dog);
        dog.eat();
        dog.bark();
        dog.sleep();

        System.out.println();

        System.out.println(cat);
        cat.eat();
        cat.purr();
        cat.sleep();
    }
}

File: code/class-04-inheritance/inheritance_demo.py
"""
inheritance_demo.py

Class 04 – Inheritance

Demonstrates:
  - Parent class with shared attributes and methods
  - Child classes using inheritance (implicit in Python via parentheses)
  - super() to call the parent __init__ and override methods
  - Method overriding with extended behaviour
"""


# ── Parent class ─────────────────────────────────────────────────────────────

class Animal:
    """Base class for all animals."""

    def __init__(self, name: str, age: int) -> None:
        self.name = name
        self.age  = age

    def eat(self) -> None:
        print(f"{self.name} is eating.")

    def sleep(self) -> None:
        print(f"{self.name} is sleeping.")

    def __str__(self) -> str:
        return f"Animal[name={self.name}, age={self.age}]"


# ── Dog (child) ───────────────────────────────────────────────────────────────

class Dog(Animal):
    """A dog that extends Animal."""

    def __init__(self, name: str, age: int, breed: str) -> None:
        super().__init__(name, age)   # call Animal.__init__
        self.breed = breed

    def bark(self) -> None:
        print(f"{self.name} says: Woof!")

    def eat(self) -> None:          # override
        super().eat()               # reuse parent behaviour …
        print(f"{self.name} wags its tail while eating.")   # … then extend it

    def __str__(self) -> str:
        return f"Dog[name={self.name}, age={self.age}, breed={self.breed}]"


# ── Cat (child) ───────────────────────────────────────────────────────────────

class Cat(Animal):
    """A cat that extends Animal."""

    def __init__(self, name: str, age: int, is_indoor: bool = True) -> None:
        super().__init__(name, age)
        self.is_indoor = is_indoor

    def purr(self) -> None:
        print(f"{self.name} purrs contentedly.")

    def eat(self) -> None:          # override
        print(f"{self.name} eats very delicately.")

    def __str__(self) -> str:
        return f"Cat[name={self.name}, age={self.age}, indoor={self.is_indoor}]"


# ── Demo ─────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    dog = Dog("Rex",      3, "Labrador")
    cat = Cat("Whiskers", 5, is_indoor=True)

    print(dog)
    dog.eat()
    dog.bark()
    dog.sleep()

    print()

    print(cat)
    cat.eat()
    cat.purr()
    cat.sleep()

    # isinstance checks
    print()
    print(f"dog is Animal? {isinstance(dog, Animal)}")   # True
    print(f"cat is Dog?    {isinstance(cat, Dog)}")      # False
